x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authc/jwt/JwtUtil.java

```diff
@@ -0,0 +1,884 @@
+                final int proxyPort = realmConfig.getSetting(JwtRealmSettings.HTTP_PROXY_PORT);
+
+                final RequestConfig.Builder requestConfigBuilder = RequestConfig.custom();
+                requestConfigBuilder.setConnectTimeout(tcpConnectTimeoutMillis)
+                    .setConnectionRequestTimeout(tcpConnectionReadTimeoutSec)
+                    .setSocketTimeout(tcpSocketTimeout);
+                if (Strings.hasText(proxyAddress) && Strings.hasText(proxyScheme)) {
+                    requestConfigBuilder.setProxy(new HttpHost(proxyAddress, proxyPort, proxyScheme));
+                }
+                final RegistryBuilder<SchemeIOSessionStrategy> sessionStrategyBuilder = RegistryBuilder.<SchemeIOSessionStrategy>create();
+                //// final String realmConfigPrefixSslSettings = RealmSettings.realmSslPrefix(realmIdentifier);
+                // final SSLContext sslContext = sslService.sslContext(elasticsearchSslConfig);
+                // final HostnameVerifier hostnameVerifier = SSLService.getHostnameVerifier(elasticsearchSslConfig);
+                // final SSLIOSessionStrategy sslIOSessionStrategy = new SSLIOSessionStrategy(sslContext, hostnameVerifier);
+                final SSLIOSessionStrategy sslIOSessionStrategy = sslService.sslIOSessionStrategy(elasticsearchSslConfig);
+                sessionStrategyBuilder.register("https", sslIOSessionStrategy);
+
+                final PoolingNHttpClientConnectionManager httpClientConnectionManager = new PoolingNHttpClientConnectionManager(
+                    new DefaultConnectingIOReactor(),
+                    sessionStrategyBuilder.build()
+                );
+                httpClientConnectionManager.setDefaultMaxPerRoute(httpMaxEndpointConnections);
+                httpClientConnectionManager.setMaxTotal(httpMaxConnections);
+
+                final CloseableHttpAsyncClient httpAsyncClient = HttpAsyncClients.custom()
+                    .setConnectionManager(httpClientConnectionManager)
+                    .setDefaultRequestConfig(requestConfigBuilder.build())
+                    .build();
+                httpAsyncClient.start();
+                return httpAsyncClient;
+            });
+        } catch (PrivilegedActionException e) {
+            throw new IllegalStateException("Unable to create a HttpAsyncClient instance", e);
+        }
+    }
+
+    /**
+     * Use the HTTP Client to get URL content bytes up to N max bytes.
+     * @param httpClient Configured HTTP/HTTPS client.
+     * @param uri URI to download.
+     * @param maxBytes Max bytes to read for the URI. Use Integer.MAX_VALUE to read all.
+     * @return Byte array of the URI contents up to N max bytes.
+     * @throws Exception Security exception or HTTP/HTTPS failure exception.
+     */
+    @SuppressWarnings({ "removal", "unusedThrown" })
+    public static byte[] readBytes(final CloseableHttpAsyncClient httpClient, final URI uri, final int maxBytes) throws Exception {
```

Comment:
It mirrors the Files.readBytes() which takes a byte count.

Also, I do foresee the possibility of using this to read 1 byte. That can be used to check if a URL exists, contents are non-empty, or TLS settings are valid, without downloading the full contents.

Reply:
I personally see neither of those as a good enough reason to add this now. I suggest we remove it since it is unused, and if and when we actually need to do something like (check if a URL exists, contents are non-empty, or TLS settings are valid)  we can see if this is the way to do it.

Reply:
Removed exception. Removed maxBytes.